File: tests/src/nl/t64/game/rpg/components/loot/LootTest.java
package nl.t64.game.rpg.components.loot;

import nl.t64.game.rpg.GameTest;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.entry;


class LootTest extends GameTest {

    private LootContainer lootContainer;

    @BeforeEach
    private void setup() {
        lootContainer = new LootContainer();
    }

    @Test
    void whenSparklesAreCreated_ShouldHaveCorrectDataInside() {
        Loot sparkle = lootContainer.getLoot("sparkle0001");
        Map<String, Integer> expected = Map.of("gemstone", 1);
        assertThat(sparkle.isTaken()).isFalse();
        assertThat(sparkle.getContent()).hasSize(1);
        Map.Entry<String, Integer> entry = expected.entrySet().iterator().next();
        assertThat(sparkle.getContent()).containsOnly(entry(entry.getKey(), entry.getValue()));
    }

    @Test
    void whenChestsAreCreated_ShouldHaveCorrectDataInside() {
        Loot chest = lootContainer.getLoot("chest0001");
        Map<String, Integer> expected = Map.of("healing_potion", 3);
        assertThat(chest.isTaken()).isFalse();
        assertThat(chest.isTrapped()).isFalse();
        assertThat(chest.isLocked()).isFalse();
        assertThat(chest.getContent()).hasSize(1);
        Map.Entry<String, Integer> entry = expected.entrySet().iterator().next();
        assertThat(chest.getContent()).containsOnly(entry(entry.getKey(), entry.getValue()));
    }

    @Test
    void whenQuestsAreCreated_ShouldHaveCorrectDataInside() {
        Loot reward = lootContainer.getLoot("quest0001");
        assertThat(reward.isTaken()).isFalse();
        assertThat(reward.getContent()).hasSize(1);
        assertThat(reward.getXp()).isEqualTo(2);
        reward.clearXp();
        assertThat(reward.getContent()).hasSize(1);
        assertThat(reward.getXp()).isZero();
    }

    @Test
    void whenLootIdDoesNotExistLikeInQuestWithNoReward_ShouldReturnEmptyReward() {
        Loot reward = lootContainer.getLoot("doesNotExist");
        assertThat(reward.getContent()).isEmpty();
        assertThat(reward.getXp()).isZero();
        assertThat(reward.isTaken()).isTrue();
        assertThat(reward.isTrapped()).isFalse();
        assertThat(reward.canDisarmTrap(0)).isTrue();
        assertThat(reward.isLocked()).isFalse();
        assertThat(reward.canPickLock(0)).isTrue();
    }

    @Test
    void whenChestsHasTrapOrLock_ShouldNeedToDisarmOrPick() {
        int thieflevel = 8;
        int mechaniclevel = 8;
        Loot chest = lootContainer.getLoot("chest0005");
        assertThat(chest.isTaken()).isFalse();
        assertThat(chest.isTrapped()).isTrue();
        assertThat(chest.getTrapLevel()).isEqualTo(6);
        assertThat(chest.canDisarmTrap(mechaniclevel)).isTrue();
        assertThat(chest.isLocked()).isTrue();
        assertThat(chest.getLockLevel()).isEqualTo(10);
        assertThat(chest.canPickLock(thieflevel)).isFalse();
        assertThat(chest.getContent()).hasSize(1);
        assertThat(chest.getContent()).isEqualTo(Map.of("gold", 1000));
        chest.disarmTrap();
        chest.pickLock();
        chest.clearContent();
        assertThat(chest.isTaken()).isTrue();
        assertThat(chest.isTrapped()).isFalse();
        assertThat(chest.getTrapLevel()).isZero();
        assertThat(chest.canDisarmTrap(mechaniclevel)).isTrue();
        assertThat(chest.isLocked()).isFalse();
        assertThat(chest.getLockLevel()).isZero();
        assertThat(chest.canPickLock(thieflevel)).isTrue();
        assertThat(chest.getContent()).isEmpty();
        chest.updateContent(Map.of("herb", 10));
        assertThat(chest.isTaken()).isFalse();
        assertThat(chest.isTrapped()).isFalse();
        assertThat(chest.getTrapLevel()).isZero();
        assertThat(chest.canDisarmTrap(mechaniclevel)).isTrue();
        assertThat(chest.isLocked()).isFalse();
        assertThat(chest.getLockLevel()).isZero();
        assertThat(chest.canPickLock(thieflevel)).isTrue();
        assertThat(chest.getContent()).hasSize(1);
    }

    @Test
    void whenQuestHasBonusRemoved_ShouldRemoveBonus() {
        Loot reward = lootContainer.getLoot("quest0006");
        assertThat(reward.isTaken()).isFalse();
        assertThat(reward.getContent()).hasSize(3);
        assertThat(reward.getContent()).isEqualTo(Map.of("gold", 1, "bonus_gold", 1, "bonus_herb", 1));
        assertThat(reward.getXp()).isEqualTo(2);
        reward.removeBonus();
        assertThat(reward.getContent()).hasSize(1);
        assertThat(reward.getContent()).isEqualTo(Map.of("gold", 1));
    }

    @Test
    void whenQuestHasBonusHandled_ShouldMergeAndRemoveBonus() {
        Loot reward = lootContainer.getLoot("quest0006");
        assertThat(reward.isTaken()).isFalse();
        assertThat(reward.getContent()).hasSize(3);
        assertThat(reward.getContent()).isEqualTo(Map.of("gold", 1, "bonus_gold", 1, "bonus_herb", 1));
        assertThat(reward.getXp()).isEqualTo(2);
        reward.handleBonus();
        assertThat(reward.getContent()).hasSize(2);
        assertThat(reward.getContent()).isEqualTo(Map.of("gold", 2, "herb", 1));
    }

}

File: tests/src/nl/t64/game/rpg/components/loot/LootTest.kt
package nl.t64.game.rpg.components.loot

import nl.t64.game.rpg.GameTest
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.entry
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test


internal class LootTest : GameTest() {

    private lateinit var lootContainer: LootContainer

    @BeforeEach
    private fun setup() {
        lootContainer = LootContainer()
    }

    @Test
    fun whenSparklesAreCreated_ShouldHaveCorrectDataInside() {
        val sparkle = lootContainer.getLoot("sparkle0001")
        val expected = mapOf(Pair("gemstone", 1))
        assertThat(sparkle.isTaken()).isFalse
        assertThat(sparkle.content).hasSize(1)
        val (key, value) = expected.entries.iterator().next()
        assertThat(sparkle.content).containsOnly(entry(key, value))
    }

    @Test
    fun whenChestsAreCreated_ShouldHaveCorrectDataInside() {
        val chest = lootContainer.getLoot("chest0001")
        val expected = mapOf(Pair("healing_potion", 3))
        assertThat(chest.isTaken()).isFalse
        assertThat(chest.isTrapped()).isFalse
        assertThat(chest.isLocked()).isFalse
        assertThat(chest.content).hasSize(1)
        val (key, value) = expected.entries.iterator().next()
        assertThat(chest.content).containsOnly(entry(key, value))
    }

    @Test
    fun whenQuestsAreCreated_ShouldHaveCorrectDataInside() {
        val reward = lootContainer.getLoot("quest0001")
        assertThat(reward.isTaken()).isFalse
        assertThat(reward.content).hasSize(1)
        assertThat(reward.xp).isEqualTo(2)
        reward.clearXp()
        assertThat(reward.content).hasSize(1)
        assertThat(reward.xp).isZero
    }

    @Test
    fun whenLootIdDoesNotExistLikeInQuestWithNoReward_ShouldReturnEmptyReward() {
        val reward = lootContainer.getLoot("doesNotExist")
        assertThat(reward.content).isEmpty()
        assertThat(reward.xp).isZero
        assertThat(reward.isTaken()).isTrue
        assertThat(reward.isTrapped()).isFalse
        assertThat(reward.canDisarmTrap(0)).isTrue
        assertThat(reward.isLocked()).isFalse
        assertThat(reward.canPickLock(0)).isTrue
    }

    @Test
    fun whenChestsHasTrapOrLock_ShouldNeedToDisarmOrPick() {
        val thieflevel = 8
        val mechaniclevel = 8
        val chest = lootContainer.getLoot("chest0005")
        assertThat(chest.isTaken()).isFalse
        assertThat(chest.isTrapped()).isTrue
        assertThat(chest.trapLevel).isEqualTo(6)
        assertThat(chest.canDisarmTrap(mechaniclevel)).isTrue
        assertThat(chest.isLocked()).isTrue
        assertThat(chest.lockLevel).isEqualTo(10)
        assertThat(chest.canPickLock(thieflevel)).isFalse
        assertThat(chest.content).hasSize(1)
        assertThat(chest.content).isEqualTo(mapOf(Pair("gold", 1000)))
        chest.disarmTrap()
        chest.pickLock()
        chest.clearContent()
        assertThat(chest.isTaken()).isTrue
        assertThat(chest.isTrapped()).isFalse
        assertThat(chest.trapLevel).isZero
        assertThat(chest.canDisarmTrap(mechaniclevel)).isTrue
        assertThat(chest.isLocked()).isFalse
        assertThat(chest.lockLevel).isZero
        assertThat(chest.canPickLock(thieflevel)).isTrue
        assertThat(chest.content).isEmpty()
        chest.updateContent(mutableMapOf(Pair("herb", 10)))
        assertThat(chest.isTaken()).isFalse
        assertThat(chest.isTrapped()).isFalse
        assertThat(chest.trapLevel).isZero
        assertThat(chest.canDisarmTrap(mechaniclevel)).isTrue
        assertThat(chest.isLocked()).isFalse
        assertThat(chest.lockLevel).isZero
        assertThat(chest.canPickLock(thieflevel)).isTrue
        assertThat(chest.content).hasSize(1)
    }

    @Test
    fun whenQuestHasBonusRemoved_ShouldRemoveBonus() {
        val reward = lootContainer.getLoot("quest0006")
        assertThat(reward.isTaken()).isFalse
        assertThat(reward.content).hasSize(3)
        assertThat(reward.content).isEqualTo(mapOf(Pair("gold", 1), Pair("bonus_gold", 1), Pair("bonus_herb", 1)))
        assertThat(reward.xp).isEqualTo(2)
        reward.removeBonus()
        assertThat(reward.content).hasSize(1)
        assertThat(reward.content).isEqualTo(mapOf(Pair("gold", 1)))
    }

    @Test
    fun whenQuestHasBonusHandled_ShouldMergeAndRemoveBonus() {
        val reward = lootContainer.getLoot("quest0006")
        assertThat(reward.isTaken()).isFalse
        assertThat(reward.content).hasSize(3)
        assertThat(reward.content).isEqualTo(mapOf(Pair("gold", 1), Pair("bonus_gold", 1), Pair("bonus_herb", 1)))
        assertThat(reward.xp).isEqualTo(2)
        reward.handleBonus()
        assertThat(reward.content).hasSize(2)
        assertThat(reward.content).isEqualTo(mapOf(Pair("gold", 2), Pair("herb", 1)))
    }

}
